include/taskolib/GitObjectWrapper.h

```diff
@@ -0,0 +1,166 @@
+ *
+ * \copyright Copyright 2023 Deutsches Elektronen-Synchrotron (DESY), Hamburg
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation, either version 2.1 of the license, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+// SPDX-License-Identifier: LGPL-2.1-or-later
+
+#include <string>
+#include <typeinfo>
+#include <utility>
+
+#include <git2.h>
+
+
+namespace git {
+
+// free C-type pointer (overload)
+void free_libgit_ptr(git_tree* tree);
+void free_libgit_ptr(git_signature* signature);
+void free_libgit_ptr(git_index* index);
+void free_libgit_ptr(git_repository* repo);
+void free_libgit_ptr(git_remote* remote);
+void free_libgit_ptr(git_commit* commit);
+void free_libgit_ptr(git_status_list* status);
+
+
+/**
+ * Template class wrapper for all libgit2 pointer types starting with git_{a-z}*.
+ * Copy methods are excluded to prevent double ownership of said pointer.
+ * 
+ * \tparam T: A dereferencing from a pointer type like git_remote*, git_tree*, git_commit*, ...
+*/
+template <class T>
+class LibGitPointer
+{
+public:
+    /// (default) Default-construct an LibGitPointer. The managed pointer is NULL.
+    LibGitPointer(){};
+    
+    /// (construct by value) Construct a LibGitPointer via a C-pointer from libgit2.
+    LibGitPointer(T *val): val_{val} {};
+
+    /// (move constructor) Move a C-pointer from an existing LibGitPointer to a new one.
+    LibGitPointer(LibGitPointer&& lg): val_{std::exchange(lg.val_, nullptr)} {};
+
+    /// (destructor) Destruct the Object by freeing the C-type pointer with a libgit function.
+    ~LibGitPointer(){ free_libgit_ptr(val_); };
+
+    /// (move assignment) move value from other object to this object
+    LibGitPointer& operator=(LibGitPointer&& lg) noexcept
+    {
+        val_ = std::move(lg.val_);
+        lg.val_ = nullptr;
+        return *this;
+    }
+
+    /// (copy constructor) Shall not be used to prevent double ownership of pointer.
+    LibGitPointer(const LibGitPointer&) = delete;
+
+    /// (copy assignment) Shall not be used to prevent double ownership of pointer.
+    LibGitPointer& operator=(const LibGitPointer&) = delete;
+
+    /**
+     * function to free member variables of GitRepository object before deconstruction of their owner.
+     * necessary because they have to be freed before the libgit2 environment deconstruction, see:
+     * git_libgit2_shutdow()
+     */
+    void reset()
```

Comment:
Hmm ... when we called `reset()` we destructor the internal pointer. Isn't it better to remove `reset()` and destruct `val_` in the destructor?

Reply:
This is also symmetric with `std::unique_ptr`, which has a similar `reset()`.